Patch-Source: https://lists.sr.ht/~mpu/qbe/patches/54249
---
From nobody Tue Nov 26 14:47:04 2024
Authentication-Results: mail-a.sr.ht; dkim=pass header.d=disroot.org header.i=@disroot.org
Received: from layka.disroot.org (layka.disroot.org [178.21.23.139])
	by mail-a.sr.ht (Postfix) with ESMTPS id 177E220338
	for <~mpu/qbe@lists.sr.ht>; Fri,  2 Aug 2024 11:40:12 +0000 (UTC)
Received: from localhost (localhost [127.0.0.1])
	by disroot.org (Postfix) with ESMTP id D58EE4157B;
	Fri,  2 Aug 2024 13:40:11 +0200 (CEST)
X-Virus-Scanned: SPAM Filter at disroot.org
Received: from layka.disroot.org ([127.0.0.1])
	by localhost (disroot.org [127.0.0.1]) (amavisd-new, port 10024)
	with ESMTP id Vw9C0KfyJUMT; Fri,  2 Aug 2024 13:40:11 +0200 (CEST)
From: Alexey Yerin <yyp@disroot.org>
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/simple; d=disroot.org; s=mail;
	t=1722598811; bh=NXLwwzVyqkGcgVXTMp7liB4yBx+rjBn7W+KzqVdahpU=;
	h=From:To:Cc:Subject:Date;
	b=JjieT9AN6Qb9OJlC7ATRKHNjIlj4LjEKRQlfXo0Q18nWNk+znpIyE2aTkISKZYyrc
	 un2//NSd5o/yJFAOK6fPxWU6os7m5txiZ06iV13nClJ42pfjIOKFRyxtfJ/mzdjm50
	 0SwEBdDbZrcSlKrmwEsmIUDja+w49AfetoQL1dhyQPM2d5TV43Pt/T38Iab/8SuR1i
	 ZVFLS6Ip7IBu9UTtQuU7fVFM03JTHhO5S6LujRVikRnI027gNBT76/PPVgldZHhWx7
	 QSTi8GrqKOh1L2kn0uuSwS6N+MG+Xzp96TO1oD66bQ2jVSdufQrqvQdUNqk8dNghGa
	 w1ck5nI2I4pig==
To: ~mpu/qbe@lists.sr.ht
Cc: Alexey Yerin <yyp@disroot.org>
Subject: [PATCH] arm64/isel: Avoid signed overflow when handling immediates
Date: Fri,  2 Aug 2024 14:39:07 +0300
Message-ID: <20240802113942.14588-1-yyp@disroot.org>
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit

Clang incorrectly optimizes this negation with -O2 and causes QBE to
emit 0 in place of INT64_MIN.
---
 arm64/isel.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arm64/isel.c b/arm64/isel.c
index 062beb3..9ce6adc 100644
--- a/arm64/isel.c
+++ b/arm64/isel.c
@@ -24,7 +24,7 @@ imm(Con *c, int k, int64_t *pn)
 	i = Iplo12;
 	if (n < 0) {
 		i = Inlo12;
-		n = -n;
+		n = -(uint64_t)n;
 	}
 	*pn = n;
 	if ((n & 0x000fff) == n)
-- 
2.46.0


